File: implementations/c/bfs/bfs_mid.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#define N 1000

static bool adj[N][N];
static bool visited[N];
static int  queue[N];

static void setup(void) {
    memset(adj, false, sizeof(adj));
    for (int i = 0; i < N - 1; i++) {
        adj[i][i + 1] = true;
        adj[i + 1][i] = true;
    }
    for (int i = 0; i < N; i += 10) {
        int j = (i * 7 + 3) % N;
        adj[i][j] = true;
        adj[j][i] = true;
    }
}

static int bfs(void) {
    memset(visited, false, sizeof(visited));
    int head = 0, tail = 0, count = 0;

    visited[0] = true;
    queue[tail++] = 0;

    while (head < tail) {
        int cur = queue[head++];
        count++;
        for (int nb = 0; nb < N; nb++) {
            if (adj[cur][nb] && !visited[nb]) {
                visited[nb]   = true;
                queue[tail++] = nb;
            }
        }
    }
    return count;
}

int main(void) {
    setup();

    printf("ready\n");
    fflush(stdout);

    char buf[64];
    while (fgets(buf, sizeof(buf), stdin)) {
        printf("%d\n", bfs());
        fflush(stdout);
    }

    return 0;
}

File: implementations/go/bfs/mid/bfs_mid.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const N = 1000

var adj [N][N]bool
var visited [N]bool
var queue [N]int

func setup() {
	for i := 0; i < N-1; i++ {
		adj[i][i+1] = true
		adj[i+1][i] = true
	}
	for i := 0; i < N; i += 10 {
		j := (i*7 + 3) % N
		adj[i][j] = true
		adj[j][i] = true
	}
}

func bfs() int {
	for i := 0; i < N; i++ {
		visited[i] = false
	}
	head, tail, count := 0, 0, 0
	visited[0] = true
	queue[tail] = 0
	tail++
	for head < tail {
		cur := queue[head]
		head++
		count++
		for nb := 0; nb < N; nb++ {
			if adj[cur][nb] && !visited[nb] {
				visited[nb] = true
				queue[tail] = nb
				tail++
			}
		}
	}
	return count
}

func main() {
	setup()

	fmt.Println("ready")

	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		fmt.Println(bfs())
	}
}
